### ISSUE TITLE
[Platform] Use systemd rather than cron for liveness checks

### ISSUE THREAD

**Issue by ramkumarvs** [yugabyte/yugabyte-db#4123]
One of the users brought this up, right now we use cron job for our liveness checks, make be it makes sense to use systemd thought?

@bmatican @mbautin

Aha! Link: https://yugabyte-test.aha.io/features/PLATFORM-545

**Comment by rkarthik007** [comment]
@ramkumarvs - `cron` can be installed without `sudo` permissions, while that's not the case with `systemd`. IIRC we switched from the latter to the former for this purpose.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4123-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4123 — [Platform] Use systemd rather than cron for liveness checks

Kind: feature request (area/platform, YBA). Filed 2020-03-31 by YB engineering
(Aha PLATFORM-545). Validation mode: behavior-delta (filing-time state vs
requested state vs current state), per protocol for feature requests — no crash
signature exists.

## Verdict
NOT-REPRODUCED — the premise ("right now we use cron job for our liveness
checks") is no longer true. The requested systemd-based liveness shipped on
2021-07-30 (commit `7eec11755572`, "[#9236] Replace cron jobs with systemd
services for yb-master, yb-tserver, clean_cores, and zip_purge_yb_logs"), a
cron→systemd migration task followed ([#9237]), and since 2025-08-05
(`72a231d`, PLAT-17993) systemd is the default AND the only allowed mode for
new universes. Issue #6174 — the same core request — was closed COMPLETED
2021-07-30, 11 minutes after the delivery commit; a 2021-10-07 comment on
#4123 already notes "Dup (solved) in #6174". Quick verdict: Fixed.

## Root cause
Filing-time state (feature-request baseline), at repo ref `769b2a4fcbf`
(2020-03-31T20:16:53Z, 24 minutes before filing):

- `managed/devops/roles/configure-cluster-server/tasks/main.yml:134-140` —
  ansible task "Setup cron to check liveness of {{ yb_process_type }}"
  installs a per-minute user crontab entry:
  `yb-server-ctl.sh <process> cron-check || yb-server-ctl.sh <process> start`.
- `managed/devops/roles/configure-cluster-server/templates/yb-server-ctl.sh.j2:226-233`
  — the hidden `cron-check` subcommand implementing the liveness probe
  (pid-file check; cron restarts the daemon when it fails).
- No `.service`/`.timer` template existed anywhere under
  `managed/devops/roles/` at that ref (templates dir held only
  `clean_cores.sh.j2`, `yb-server-ctl.sh.j2`, `zip_purge_yb_logs.sh.j2`).

A same-day maintainer comment (rkarthik007) explains the historical choice:
cron works without sudo; systemd (as then conceived, system-level units) does
not — YBA had deliberately switched systemd→cron earlier.

Current state (master, checkout `46a9f7adfd`):

- `managed/node-agent/resources/templates/server/yb-master.service:17-18` —
  `Restart=always` / `RestartSec=5`: systemd supervision IS the liveness
  mechanism (line 19 comment: "matches existing cron behavior"). Same for
  `yb-tserver.service`, `yb-controller.service`; the auxiliary cron jobs
  became `.timer` units (`yb-clean_cores.timer`, `yb-zip_purge_yb_logs.timer`,
  `yb-collect_metrics.timer`).
- `managed/src/main/java/com/yugabyte/yw/forms/UniverseDefinitionTaskParams.java:1384`
  — `public Boolean useSystemd = true;` (default).
- `managed/src/main/java/com/yugabyte/yw/controllers/UniverseClustersController.java:82-83`
  — null `useSystemd` forced to `true` on universe create.
- `managed/src/main/java/com/yugabyte/yw/controllers/handlers/UniverseCRUDHandler.java:446-452`
  — new primary-cluster creation with `useSystemd=false` is rejected:
  "Only systemd is supported. Set useSystemd=true for the primary cluster".
- The sudo objection is resolved via USER-level systemd:
  `managed/node-agent/ynp/cmd/main.go:343-344` defaults
  `use_system_level_systemd=false`, and
  `managed/node-agent/app/task/server_control.go:85` runs
  `loginctl enable-linger <user>` so user units survive logout; sudo is needed
  only at provision time.

Master "affected": no — the requested behavior is implemented and enforced.
No fix PR is warranted; the issue itself should be closed as
duplicate-of/completed-by #6174.

## Trigger matrix
Behavior-delta matrix; each cell predicted from the pattern hint before
inspection, then confirmed against immutable refs (GitHub contents/commits
API — local checkout is shallow) or the current tree.

| Condition | Predicted | Observed |
|---|---|---|
| Liveness mechanism at filing (2020-03-31, ref `769b2a4fcbf`) | cron-only; crontab entries in devops ansible | CONFIRMED — per-minute crontab `cron-check \|\| start` (`configure-cluster-server/tasks/main.yml:134-140`); `cron-check` in `yb-server-ctl.sh.j2:226`; zero `.service` templates in devops roles |
| systemd delivery | shipped later: unit files + timers replacing cron | CONFIRMED — `7eec11755572` (2021-07-30) adds `provision-cluster-server/templates/yb-{master,tserver}.service` + `yb-clean_cores.timer`/`yb-zip_purge_yb_logs.timer`; #6174 closed COMPLETED 2021-07-30T17:07:37Z, 11 min after |
| Migration path for existing cron universes | dedicated upgrade task | CONFIRMED — `9495cefa12ea` (2021-08-11) "[#9237] Upgrade cron based universes to systemd universes"; `SystemdUpgrade.java:85-107` on master; croncheck disabled for systemd universes (`42a45101a305`, 2021-08-12) |
| Default for new universes today | systemd is default path | CONFIRMED (stronger than predicted) — default `true` (`UniverseDefinitionTaskParams.java:1384`) AND `useSystemd=false` rejected at create (`UniverseCRUDHandler.java:446-452`), both from `72a231d` (2025-08-05, PLAT-17993) |
| Liveness on master | systemd `Restart=` supervision | CONFIRMED — `yb-master.service:17-18` `Restart=always`/`RestartSec=5`; comment "[matches existing cron behavior]" |
| 2020 sudo objection | addressed by user-level systemd | CONFIRMED — `use_system_level_systemd` defaults false (`ynp/cmd/main.go:343-344`, `resources/ynp/configs/config.j2:113`); `loginctl enable-linger` (`server_control.go:85`, `configure_os/templates/run.j2:242`) |
| Residual cron support | legacy universes still cron until upgraded | CONFIRMED — `HealthChecker.java:1214-1216` passes `--cronbased` for non-systemd universes; `AnsibleConfigureServers.java:315` retains `!useSystemd` branch; ansible cron path removed from YBA image only in `305866f4e878` (2026-04-29, PLAT-19353) |

## Reproduction
Executable repro N/A — this is a feature request, not a crash/wrong-result
bug; there is no failure signature to grep for, and validating "YBA uses cron
vs systemd for liveness" live would require provisioning a full YBA control
plane plus a cloud universe (out of scope for a validation wave). The
behavior-delta is instead proven from immutable public artifacts,
independently re-runnable:

```sh
# 1. Filing-era SHA (24 min before the issue was filed):
gh api "repos/yugabyte/yugabyte-db/commits?until=2020-03-31T20:40:06Z&per_page=1" \
  --jq '.[0].sha'   # -> 769b2a4fcbf434f9486fa79191998d2d320ba570

# 2. Filing-era cron liveness (the "root cause" state):
gh api "repos/yugabyte/yugabyte-db/contents/managed/devops/roles/configure-cluster-server/tasks/main.yml?ref=769b2a4fcbf434f9486fa79191998d2d320ba570" \
  --jq '.content' | base64 -d | grep -n -A6 "Setup cron to check liveness"

# 3. Delivery commit (earliest commit touching the systemd unit template):
gh api --paginate "repos/yugabyte/yugabyte-db/commits?path=managed/devops/roles/provision-cluster-server/templates/yb-master.service&per_page=100" \
  --jq '.[] | .sha[0:12] + " " + .commit.committer.date + " " + (.commit.message | split("\n")[0])' | tail -1
# -> 7eec11755572 2021-07-30T16:56:14Z [#9236] Replace cron jobs with systemd services ...

# 4. Current enforcement (master):
grep -n "Only systemd is supported" \
  managed/src/main/java/com/yugabyte/yw/controllers/handlers/UniverseCRUDHandler.java
grep -n "Restart=always" \
  managed/node-agent/resources/templates/server/yb-master.service
```

Era snapshots preserved: `/tmp/claude/yb-issue-4123-era-configure-tasks-main.yml`
(filing-era ansible tasks; cron liveness at lines 134-140) and
`/tmp/claude/yb-issue-4123-era-yb-server-ctl.sh.j2` (filing-era control
script; `cron-check` at line 226).

## Versions tested
- Filing-era tree: ref `769b2a4fcbf434f9486fa79191998d2d320ba570`
  (2020-03-31T20:16:53Z) via GitHub contents API — cron-only.
- Delivery ref: `7eec11755572f7b36f75df70923cd69e68cdb7ee`
  (2021-07-30T16:56:14Z) — first systemd unit templates; #6174 closed against
  it.
- Current master: local checkout `46a9f7adfd226d7bc3315c38b660eef69bb11196`
  (2026-08) — systemd default + enforced; cron retained only as legacy-universe
  runtime residual.
- Master affected: no (request delivered).

## Proposed fix direction
No code fix. Housekeeping: close #4123 as completed/duplicate of #6174
(closed COMPLETED 2021-07-30), citing delivery lineage `7eec11755572`
([#9236], initial systemd services), `9495cefa12ea` ([#9237], cron→systemd
upgrade task), and `72a231d` (PLAT-17993, systemd default + cron-create
rejection). Residuals a maintainer may still care about (none block closure):
legacy pre-2021 universes remain on cron until a SystemdUpgrade run
(`SystemdUpgrade.java:85-107`), and `HealthChecker.java:1214-1216` /
`AnsibleConfigureServers.java:315` keep the `--cronbased`/`!useSystemd`
compatibility branches alive for them.

## Test strategy
N/A for a fix (none proposed). The shipped behavior is already covered by
YBA unit tests exercising the systemd path — e.g.
`managed/src/test/java/com/yugabyte/yw/commissioner/tasks/upgrade/SystemdUpgradeTest.java`
(cron→systemd migration task) and NodeManager command-generation tests
asserting systemd server-control args
(`managed/src/test/java/com/yugabyte/yw/common/NodeManagerTest.java`). Any
future removal of the legacy cron branches should extend those two suites.

## Confidentiality check
Clean. All content derives from the public GitHub issue (#4123, #6174), the
public yugabyte/yugabyte-db repository (public SHAs, public file paths, public
commit messages), and public Jira/PLAT ticket IDs already embedded in public
commit messages. No customer identifiers, PII, schemas, secrets, or internal
hostnames/URLs. The Aha link (PLATFORM-545) appears in the public issue body
itself.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4123-era-configure-tasks-main.yml.txt`):

# Copyright 2019 YugaByte, Inc. and Contributors
#
# Licensed under the Polyform Free Trial License 1.0.0 (the "License"); you
# may not use this file except in compliance with the License. You
# may obtain a copy of the License at
#
# https://github.com/YugaByte/yugabyte-db/blob/master/licenses/POLYFORM-FREE-TRIAL-LICENSE-1.0.0.txt

- assert:
    that:
      - yb_process_type is defined
      - user_name is defined
      - placement_cloud is defined
      - placement_region is defined
      - placement_zone is defined
      - instance_type is defined
      - mount_points is defined

- name: Setup | Create config directory for {{ yb_process_type }}
  file:
    path: "{{ yb_conf_dir }}"
    state: directory
    mode: 0755
    owner: "{{ user_name }}"
    group: "{{ user_name }}"

- name: Setup | Create bin directory for {{ user_name }}
  file:
    path: "{{ yb_bin_dir }}"
    state: directory
    mode: 0755
    owner: "{{ user_name }}"
    group: "{{ user_name }}"

- set_fact:
    _mount_points: "{{ mount_points.split(',') }}"

- set_fact:
    mount_logs_dir: "{{ _mount_points[0] }}/yb-data/{{ yb_process_type }}/logs"
    mount_cores_dir: "{{ _mount_points[0] }}/cores"

- name: Setup | Create {{ yb_process_type }} logs directory
  file:
    path: "{{ mount_logs_dir }}"
    state: directory
    mode: 0755
    owner: "{{ user_name }}"
    group: "{{ user_name }}"

- name: Setup | Symlink {{ yb_process_type }} logs directory
  file:
    src: "{{ mount_logs_dir }}"
    dest: "{{ yb_logs_dir }}"
    mode: 0755
    owner: "{{ user_name }}"
    group: "{{ user_name }}"
    state: link

- name: Configure | Remove {{ yb_process_type }} gflags
  file:
    path: "{{ yb_conf_file }}"
    state: absent
  when: replace_gflags is defined

- set_fact:
    _gflags: "{{ common_gflags | combine(process_specific_gflags[yb_process_type]) }}"

- set_fact:
    _gflags: "{{ _gflags | combine(dict(extra_gflags)) }}"
  when: extra_gflags is defined

- name: Configure | Set {{ yb_process_type }} gflags
  lineinfile:
    dest: "{{ yb_conf_file }}"
    regexp: "^--{{ item.key }}=(.*)"
    line: "--{{item.key}}={{item.value}}"
    create: yes
  with_dict: "{{ gflags | default(_gflags) }}"

- name: Configure | Add yb server ctl script
  vars:
    mount_paths: "{{ _mount_points | join(' ') }}"
  template:
    src: yb-server-ctl.sh.j2
    dest: "{{ yb_bin_dir }}/yb-server-ctl.sh"
    owner: "{{ user_name }}"
    mode: 0755

- name: Setup | Create directory to land core dumps
  file:
    path: "{{ mount_cores_dir }}"
    state: directory
    mode: 0755
    owner: "{{ user_name }}"
    group: "{{ user_name }}"

- name: Setup | Symlink cores directory to {{ user_name }} home directory
  file:
    src: "{{ mount_cores_dir }}"
    dest: "{{ yb_cores_dir }}"
    mode: 0755
    owner: "{{ user_name }}"
    group: "{{ user_name }}"
    state: link

- name: Configure | Add clean_cores script
  template:
    src: clean_cores.sh.j2
    dest: "{{ yb_bin_dir }}/clean_cores.sh"
    owner: "{{ user_name }}"
    mode: 0755

- name: Configure | Add purge logs script
  template:
    src: zip_purge_yb_logs.sh.j2
    dest: "{{ yb_bin_dir }}/zip_purge_yb_logs.sh"
    owner: "{{ user_name }}"
    mode: 0755

- name: Setup cron to cleanup core files.
  cron:
    name: "cleanup core files hourly"
    minute: "0"
    user: "{{ user_name }}"
    job: "{{ yb_bin_dir }}/clean_cores.sh"

- name: Setup cron to cleanup log files.
  cron:
    name: "cleanup yb log files hourly"
    minute: "5"
    user: "{{ user_name }}"
    job: "{{ yb_bin_dir }}/zip_purge_yb_logs.sh"

- name: Setup cron to check liveness of {{ yb_process_type }}
  cron:
    name: "Check liveness of {{ yb_process_type }}"
    minute: "*/1"
    user: "{{ user_name }}"
    job: >
      {{ yb_bin_dir }}/yb-server-ctl.sh {{ yb_process_type }} cron-check ||
      {{ yb_bin_dir }}/yb-server-ctl.sh {{ yb_process_type }} start

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4123-era-yb-server-ctl.sh.j2.txt`):

#!/usr/bin/env bash
#
# Copyright 2019 YugaByte, Inc. and Contributors
#
# Licensed under the Polyform Free Trial License 1.0.0 (the "License"); you
# may not use this file except in compliance with the License. You
# may obtain a copy of the License at
#
# https://github.com/YugaByte/yugabyte-db/blob/master/licenses/POLYFORM-FREE-TRIAL-LICENSE-1.0.0.txt

set -euo pipefail
# Redirect stderr to syslog.
exec 2> >(logger -t $(basename $0))

readonly MOUNT_PATHS=({{ mount_paths }})
readonly EXPECTED_USERNAME=({{ user_name }})
readonly CORES_DIR=({{ yb_cores_dir }})
{% raw %}
readonly NUM_MOUNTS=${#MOUNT_PATHS[@]}
{% endraw %}

if ! [ -f /.dockerenv ] && [ "$(whoami)" != "$EXPECTED_USERNAME" ]; then
  echo "Script must be run as user: $EXPECTED_USERNAME"
  exit -1
fi

print_help() {
  cat <<EOT
Usage: ${0##*/} <daemon> <command> OR
       ${0##*/} clean-instance
Daemons:
  master
  tserver
Commands:
  create        - Start the YB process on this node in cluster creation node (only applicable for
                  master)
  status        - Report the status of the YB process
  start         - Start the YB daemon on this node
  stop          - Stop the YB daemon on this node
  clean         - Remove all daemon data from this node
  clean-no-conf - Remove all daemon data from this node, except configurations
  clean-logs    - Remove all daemon logs
  -h, --help    - Show usage
EOT
}

check_pid_file() {
  if [[ ! -f ${daemon_pid_file} ]]; then
    print_err_out "Error: PID file does not exist: ${daemon_pid_file}, process is "\
      "probably not running"
    exit 1
  fi
}

exit_on_running() {
  if [[ $(check_running) -eq 0 ]]; then
    print_err_out "yb-$daemon already running"
    exit 0
  fi
}

NO_PID_FILE=200
# arg1 [OPTIONAL]: proc_pid -- the PID of the process, else, defaults to contents of daemon_pid_file
check_running() {
  set +e
  custom_proc_pid=${1:-}
  proc_pid=$custom_proc_pid
  if [[ -z $proc_pid ]]; then
    proc_pid=$(cat ${daemon_pid_file} 2>/dev/null)
    if [[ $? -ne 0 ]]; then
      echo $NO_PID_FILE
      return
    fi
  fi
  set -e
  kill -0 "$proc_pid" 2>/dev/null
  kill_outcome=$?
  # Workaround race condition between:
  # 1) cron checking the file exists and succeeding
  # 2) stop deleting the PID file and stopping the process
  # 3) cron then trying to kill and cat the file, failing and then restarting the daemon
  # If we searched for a PID file above and then we couldn't find a process to kill, then check if
  # the PID file still exists:
  # - No, then return 0 so we do not restart the process
  # - Yes, then default to outcome of kill command.
  if [[ $kill_outcome -ne 0 ]] && [[ -z $custom_proc_pid ]] && [[ ! -f ${daemon_pid_file} ]]; then
    echo 0
  else
    echo $kill_outcome
  fi
}

get_pid() {
  cat ${daemon_pid_file}
}

print_err_out() {
  echo $1 | tee /dev/stderr
}

# arg1: pid_to_wait_for -- the PID of the process to wait for
wait_pid() {
  pid_to_wait_for=$1
  end_time=$(($SECONDS + 10))
  while [[ $SECONDS -lt $end_time ]]; do
    if [[ $(check_running "$pid_to_wait_for") -eq 1 ]]; then
      break
    fi
    print_err_out "Waiting on PID: $pid_to_wait_for"
    sleep 1
  done
}

wait_for_dir_or_exit() {
  local dir_to_check=$1
  local end_time=$(($SECONDS + 10))
  while [[ $SECONDS -lt $end_time ]]; do
    if test -d $dir_to_check;
    then
      return
    else
      echo "Waiting for $dir_to_check dir..."
      sleep 1
    fi
  done
  # Exit if the directory never appeared.
  exit 1
}

clean_data_paths() {
  clean_conf_arg=${1:-true}

  set -x
  for (( i=0; i<NUM_MOUNTS; ++i ))
  do
    rm -rf "${MOUNT_PATHS[i]}"/yb-data/$daemon
  done

  print_err_out "Cleaning core files on `hostname`"
  # Need / at the end due to symlinks
  rm -rf "${CORES_DIR}"/

  if $clean_conf_arg;
  then
    print_err_out "Cleaning data for $daemon on `hostname`"
    rm -rf "{{ yb_home_dir }}"/$daemon
  else
    print_err_out "Cleaning data for $daemon on `hostname`, excluding configuration"
    # Also exclude the logs symlink as we need that to exist so we drop the PID file properly.
    find "{{ yb_home_dir }}/$daemon" -type l ! -path "*logs" | xargs rm
  fi
}

command=""
daemon=""
case "$1" in
  -h|--help)
    print_help
    exit 0
  ;;
  *)
esac

if [ "$1" == "clean-instance" ]; then
  print_err_out "Cleaning instance `hostname`"
  rm -rf "{{ yb_home_dir }}"/*
  exit 0
fi

daemon=${1:-}
command=${2:-}

# Check if we have a valid daemon and command.
if [ -z "$command" ] || [ -z "$daemon" ] ; then
  echo "No valid daemon/command specified."
  print_help
  exit 1
fi

case "$daemon" in
  master)
  ;;
  tserver)
    if [ "$command" == "create" ]; then
      echo "create command is not valid for tserver"
      exit 1
    fi
  ;;
  *)
    echo "Invalid Daemon: $daemon"
    print_help
    exit 1
esac

daemon_home="{{ yb_home_dir }}/$daemon"
daemon_logs_dir="$daemon_home/logs"
daemon_binary="$daemon_home/bin/yb-$daemon"
daemon_conf="$daemon_home/conf/server.conf"
daemon_pid_file="$daemon_logs_dir/yb-$daemon.pid"

case "$command" in
  create|start)
    create_option=""
    start_mode=""
    if [[ "$command" == "create" ]]; then
      create_option="--create_cluster"
      start_mode="with create_cluster option"
    fi
    exit_on_running
    print_err_out "Starting yb-$daemon on `hostname` $start_mode"
    # We need all three file descriptors closed so SSH connections running this do not hang!
    nohup ${daemon_binary} $create_option --flagfile ${daemon_conf} \
      >>"$daemon_home/${daemon}.out" 2>>"$daemon_home/${daemon}.err" </dev/null &
    # Save pid to file after waiting for log dir to show up.
    daemon_pid=$!
    wait_for_dir_or_exit $daemon_logs_dir
    echo $daemon_pid > $daemon_pid_file
  ;;
  status)
    if [[ $(check_running) -ne 0 ]]; then
      echo "yb-$daemon is not running"
      exit 1
    fi
    echo "yb-$daemon running with pid $(get_pid)"
  ;;
  cron-check)
  # Hidden option only used by crons to check status of the process. We want the following behavior
  # from crons:
  # 1. If the process was never started/created, don't try to start the process.
  # 2. If the process was explicitly stopped, don't try to to start the process.
  # As a result, this means that the cron should try to start the process only if the pid file
  # exists and the process is down. Therefore, if this check sees that we don't have a pid file, it
  # should just return 0 indicating to the cron that all is well.
    yb_status=$(check_running)
    if [[ $yb_status -eq $NO_PID_FILE ]]; then
      exit 0
    fi
    exit $yb_status
  ;;
  stop)
    print_err_out "Stopping yb-$daemon on `hostname`"
    if [[ $(check_running) -ne 0 ]]; then
      print_err_out "Process isn't running! Nothing to do"
      exit 0
    fi

    # Saving PID and deleting before KILL, to prevent race conditions with cron-check.
    daemon_pid=$(cat ${daemon_pid_file})
    rm ${daemon_pid_file}
    kill -TERM "$daemon_pid"
    wait_pid "$daemon_pid"

    if [[ $(check_running) -eq 0 ]]; then
      print_err_out "yb-$daemon is still running, using SIGKILL"
      kill -KILL "$daemon_pid"
    fi
  ;;
  clean)
    print_err_out "Cleaning all mount paths for $daemon on `hostname`"
    clean_data_paths true
  ;;
  clean-no-conf)
    print_err_out "Cleaning all mount paths for $daemon on `hostname`, except configuration data"
    clean_data_paths false
  ;;
  clean-logs)
    print_err_out "Cleaning $daemon logs data on `hostname`"
    (
      set -x
      rm -rf "${daemon_logs_dir}"/*
    )
  ;;
  *)
    echo "Invalid command: $command"
    print_help
    exit 1
esac

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
